File: core/jvm/main/scala/com/webank/eggroll/core/command/CommandRouter.scala
/*
 * Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.webank.eggroll.core.command

import java.lang.reflect.Method

import com.google.protobuf.ByteString
import com.webank.eggroll.core.constant.StringConstants
import com.webank.eggroll.core.meta.MetaModelPbSerdes._
import com.webank.eggroll.core.meta.{ErJob, ErTask, Meta}
import org.apache.commons.lang3.StringUtils
import org.apache.commons.lang3.reflect.{ConstructorUtils, MethodUtils}

import scala.collection.concurrent.TrieMap
import scala.collection.mutable

trait CommandRoutable {

}

object Scope {

}

object CommandRouter {
  private val serviceRouteTable = TrieMap[String, (Any, Method)]()
  private val messageParserMethodCache = mutable.Map[Class[_], Method]()

  // todo: consider different scope of target instance suck as 'singleton', 'proto', 'session' etc.
  //  This can be implemented as an annotation reader
  def register(serviceName: String,
               serviceParamTypes: List[Class[_]],
               routeToClass: Class[_] = null,
               routeToMethodName: String = null,
               routeToCallBasedClassInstance: Any = null,
               callBasedClassInstanceInitArgs: List[AnyRef] = null): Unit = this.synchronized {
    if (serviceRouteTable.contains(serviceName)) {
      throw new IllegalStateException("Service " + serviceName
        + " has been registered at: " + serviceRouteTable(serviceName))
    }

    val finalRouteToMethodName =
      if (routeToMethodName != null) routeToMethodName
      else StringUtils.substringAfterLast(serviceName, StringConstants.DOT)

    val finalRouteToClass =
      if (routeToClass != null) routeToClass
      else Class.forName(StringUtils.substringBeforeLast(serviceName, StringConstants.DOT))

    val routeToMethod = MethodUtils.getAccessibleMethod(
      finalRouteToClass, finalRouteToMethodName, serviceParamTypes: _*)

    if (routeToMethod == null) {
      throw new NoSuchMethodException("accessible method not found for " + serviceName)
    }

    val finalCallBasedInstance =
      if (routeToCallBasedClassInstance != null) {
        routeToCallBasedClassInstance
      } else {
        val finalCallBasedClassInstanceInitArgsArray =
          if (callBasedClassInstanceInitArgs != null) callBasedClassInstanceInitArgs.toArray else null
        ConstructorUtils.invokeConstructor(finalRouteToClass, finalCallBasedClassInstanceInitArgsArray: _*)
      }

    serviceRouteTable.put(serviceName, (finalCallBasedInstance, routeToMethod))
  }

  def dispatch(serviceName: String, args: Array[_ <: AnyRef], kwargs: mutable.Map[String, _ <: AnyRef]): Array[Byte] = {
    val target = query(serviceName)
    if (target == null) {
      throw new IllegalStateException("service " + serviceName + " has not been registered")
    }

    val method = target._2
    val paramTypes = method.getParameterTypes
    var paramTypeName = "unknown"

    // todo: separate to SerDes
    // deserialization
    val realArgs = args.zip(paramTypes).map {
      case (arg, paramType) => {
        if (paramType == classOf[ErTask]) {
          paramTypeName = "ErTask"
          Meta.Task.parseFrom(arg.asInstanceOf[ByteString]).fromProto()
        } else if (paramTypes == classOf[ErJob]) {
          paramTypeName = "ErJob"
          Meta.Job.parseFrom(arg.asInstanceOf[ByteString]).fromProto()
        } else {
          arg
        }
      }
    }

    // actual call
    val callResult = target._2.invoke(target._1, realArgs: _*) // method.invoke(instance, args)

    // serialization to response
    callResult match {
      case e: ErTask =>
        callResult.asInstanceOf[ErTask].toProto().toByteArray
      case e: ErJob =>
        callResult.asInstanceOf[ErJob].toProto().toByteArray
      case _ =>
        callResult.toString.getBytes()
    }

  }

  def query(serviceName: String): (Any, Method) = {
    serviceRouteTable(serviceName)
  }
}

File: python/eggroll/core/command/command_router.py
#  Copyright (c) 2019 - now, Eggroll Authors. All Rights Reserved.
#
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.

import logging
import time
from importlib import import_module

from eggroll.core.meta_model import ErTask
from eggroll.core.proto import meta_pb2
from eggroll.utils.log_utils import get_logger

L = get_logger()


class CommandRouter(object):
    __instance = None

    @staticmethod
    def get_instance():
        if not CommandRouter.__instance:
            CommandRouter()
        return CommandRouter.__instance

    def __init__(self):
        if CommandRouter.__instance:
            raise Exception(
                    'CommandRouter a singleton class and the instance has been initialized')
        else:
            CommandRouter.__instance = self
            self._service_route_table = dict()  # key: service_name, value: (instance, class, method)

    def register(self,
            service_name: str,
            service_param_deserializers: list = None,
            service_result_serializers: list = None,
            route_to_module_name: str = '',
            route_to_class_name: str = '',
            route_to_method_name: str = '',
            route_to_call_based_class_instance=None,
            call_based_class_instance_init_arg=None):
        if service_param_deserializers is None:
            service_param_deserializers = []
        if service_result_serializers is None:
            service_result_serializers = []
        if service_name in self._service_route_table:
            raise ValueError(
                    f'service {service_name} has already been registered at ${self._service_route_table[service_name]}')

        # todo:2: consider scope. now default to a 'prototype' style and no init args
        _module = import_module(route_to_module_name)
        _class = getattr(_module, route_to_class_name)
        _method = getattr(_class, route_to_method_name)

        self._service_route_table[service_name] = (None, _class, _method)
        L.info("service:{} has registered".format(service_name))

    def dispatch(self, service_name: str, args, kwargs):
        if service_name not in self._service_route_table:
            raise ValueError(f'{service_name} has not been registered yet')

        _instance, _class, _method = self._service_route_table[service_name]

        if not _instance:
            _instance = _class()

        task_name = ''
        deserialized_args = list()
        for arg in args:
            task = meta_pb2.Task()
            msg_len = task.ParseFromString(arg)
            deserialized_task = ErTask.from_proto(task)
            if not task_name:
                task_name = deserialized_task._name
            deserialized_args.append(deserialized_task)

        L.debug(f"[CS] calling: [{service_name}], task_name={task_name}, request={deserialized_args}, len={len(args)}")

        start = time.time()
        try:
            call_result = _method(_instance, *deserialized_args)
        except Exception as e:
            L.exception(f'Failed to dispatch to [{service_name}], task_name: {task_name}, request: {deserialized_args}')
            raise e
        elapsed = time.time() - start
        if L.isEnabledFor(logging.TRACE):
            L.trace(f"[CS] called (elapsed={elapsed}): [{service_name}]: task_name={task_name}, request={deserialized_args}, result={call_result}")
        else:
            L.debug(f"[CS] called (elapsed={elapsed}): [{service_name}], task_name={task_name}, request={deserialized_args}")

        # todo:2: defaulting to pb message. need changes when other types of result is present
        return [call_result.to_proto().SerializeToString()]

    def query(self, service_name: str):
        return self._service_route_table[service_name]
